### PR TITLE
Fix ticket handling, switch to exponential ticket election

### PR DESCRIPTION
**Pull request by Kubuxu**

### CODE REVIEW

File: gpbft/gpbft.go [hunk 1]
```diff
@@ -1304,10 +1299,27 @@ func (c *convergeState) Receive(sender ActorID, value ECChain, ticket Ticket, ju
 	c.senders[sender] = struct{}{}
 	key := value.Key()
 
-	// Keep only the first justification and ticket received for a value.
-	if _, found := c.values[key]; !found {
-		c.values[key] = ConvergeValue{Chain: value, Justification: justification}
-		c.tickets[key] = append(c.tickets[key], ConvergeTicket{Sender: sender, Ticket: ticket})
+	senderPower, _ := table.Get(sender)
+	// Keep only the first justification and best ticket
+	if v, found := c.values[key]; !found {
```
Comment (Stebalien) [inline gpbft/gpbft.go:1304]: Do we even need to keep multiple values? Instead, can we just keep the best ticket value (or, possibly, the best ticket for the best candidate)?

File: gpbft/gpbft.go [hunk 2]
```diff
@@ -1320,26 +1332,19 @@ func (c *convergeState) FindMaxTicketProposal(table PowerTable) ConvergeValue {
...
-		for _, ticket := range c.tickets[key] {
-			senderPower, _ := table.Get(ticket.Sender)
-			ticketHash := blake2b.Sum256(ticket.Ticket)
-			ticketAsInt := new(big.Int).SetBytes(ticketHash[:])
-			weightedTicket := new(big.Int).Mul(ticketAsInt, big.NewInt(int64(senderPower)))
-			if maxTicket == nil || weightedTicket.Cmp(maxTicket) > 0 {
-				maxTicket = weightedTicket
-				maxValue = value
-			}
+	var bestValue ConvergeValue
```
Comment (Stebalien) [inline gpbft/gpbft.go:1335]: We can simplify this logic by initializing this value to our own value. Then we can skip the `c.HasSelfValue` part.

**Review by Stebalien — APPROVED** [review]
The float logic looks correct (although a bit complex).

File: gpbft/gpbft.go (hunk 1 above)
Comment (Kubuxu) [reply gpbft/gpbft.go:1304]: We could, but that will change if we do the audit recommendation.

File: gpbft/gpbft.go (hunk 2 above)
Comment (Kubuxu) [reply gpbft/gpbft.go:1338]: I removed the self-value and did some cleanup; please review the last commit.

File: gpbft/gpbft.go (hunk 1 above)
Comment (Stebalien) [reply gpbft/gpbft.go:1313]: Wasn't the audit recommendation to skip to the second-best ticket if the best ticket doesn't have a valid candidate? In that case, we can filter candidates up-front. I.e.:

1. Ignore all CONVERGE messages with values that aren't in our candidate set.
2. Take the value with the best ticket from the remaining values.

Comment (Kubuxu) [reply gpbft/gpbft.go:1313]: yeah, I guess that could work

**Review by Stebalien — APPROVED** [review]

**Comment by Stebalien** [comment]
(happy to iterate in a new PR)

File: gpbft/gpbft.go (hunk 1 above)
Comment (masih) [reply gpbft/gpbft.go:1313]: Quick sanity check: Has the audit recommendation been implemented in this PR?

Comment (Kubuxu) [reply gpbft/gpbft.go:1313]: No it has not

**Pull request merged**